Add sample loading a GLTF file (along with altimeter in GLTF format)

Keenuts opened the pull request:
This is a sample showcasing GLTF file loading.
In its current state, this sample can also load intel's sponza (with some patching to replace png textures with DDS).
There are a lot of things to improve, this is just a baseline so we can build on top and improve it.

Main issues:
 - shaders are compiled multiple times, should not.
 - lot of assumption about the GLTF file (requires exported tangent, at least 1 material, only DDS, no GLB, no embedded GLTF.
 - PBR shader is not accurate. It's mostly so we can visualize the GLTF file. Might want to reuse the PBR shader from the mtl sample.
 - Cannot load GLB/embedded GLTF (need to add texture-from-buffer logic in BigWheels).
 - We copy each bufferview to the GPU as a "buffer". In fact we should have one memory copied from the CPU to the GPU and map Buffers on top. Need to find out how to do that with BigWheels.
 - many more

vettoreldaniele commented:
> Cannot load GLB/embedded GLTF (need to add texture-from-buffer logic in BigWheels).

Does this not work? https://github.com/google/bigwheels/blob/main/include/ppx/grfx/grfx_queue.h#LL84C30-L84C30

chaoticbob commented:
> This might, also adding the function to create a grfx::Image from a memory area is trivial.

Assuming that the dimensions, stride, and format are known ahead of time - there's already a function that does this:
https://github.com/google/bigwheels/blob/main/include/ppx/graphics_util.h#L86

There's a create method that takes an external storage pointer and doesn't copy:
https://github.com/google/bigwheels/blob/main/include/ppx/bitmap.h#L70

Keenuts commented:
> > This might, also adding the function to create a grfx::Image from a memory area is trivial.
> 
> Assuming that the dimensions, stride, and format are known ahead of time - there's already a function that does this: https://github.com/google/bigwheels/blob/main/include/ppx/graphics_util.h#L86
> 
> There's a Bitmap create method that takes an external storage pointer and doesn't copy: https://github.com/google/bigwheels/blob/main/include/ppx/bitmap.h#L70

Thanks for the pointer. When I added this code I used the following logic:
```cpp
gli::texture image = gli::load_dds(glb_bye_array, size);
ppxres = CreateImageFromCompressedImage(pQueue, image, ppImage, options);
```
The main reason I dropped the change for now is I cannot create a GLB/embedded GLTF file with the MSFT_texture_dds extension :/ It's a draft, so blender doen't support it, neither cgltf exported, outside of one binary MSFT provides for windows...

Keenuts commented on projects/28_gltf/main.cpp:
```diff
@@ -0,0 +1,774 @@
+        poolCreateInfo.uniformBuffer                  = 512;
+        poolCreateInfo.sampledImage                   = 512;
+        poolCreateInfo.sampler                        = 512;
+        poolCreateInfo.combinedImageSampler           = 512;
```
Need to find out how to replace that for DX11/DX12.

chaoticbob replied:
Is that not working on DX12?

Keenuts replied:
The combinedImageSampler is marked as unsupported (https://github.com/google/bigwheels/blob/main/src/ppx/grfx/dx12/dx12_descriptor.cpp#L44)
Actually DX11 and VK are the only 2 APIs running this sample 😅

Keenuts replied:
Removing it fails in another location I'm fixing right now (compressed textures are created with a size of 1 or 2, while DX12 requires 4 min). Don't know if I can run the DX12 sample with sponza without that line yet.

Keenuts replied:
So same issue as VK on DX12: pool runs out of descriptors of this type, but I cannot set a larger size than the default one.
Might need to fix this.

Keenuts replied:
No, just this HLSL shader. I'm not even sure why storageImage + combinedImageSampler is needed. Removing one of the other causes a pool exhaustion. Now that it is a blocker I have a reason to look into it..

Keenuts replied:
So turns out, It might be driver-related:
 - on linux, cannot run without setting both COMBINED_IMAGE_SAMPLER & STORAGE_IMAGE to 512
 - on windows (cloudtop), can run the VK version without those 2 lines.

Keenuts replied:
Found the reason:
 - radv doesn't handle a per-descriptor pool size, but allocates more or less bytes in a single pool:
 https://gitlab.freedesktop.org/mesa/mesa/-/blob/main/src/amd/vulkan/radv_descriptor_set.c#L908
 So the real issue was that we allocated more than 512 descriptors for samplers/sampled_image. But it was fine because we allocated less than 512 uniform buffers. Until we reached this limit.
 Adding any other type bumps this global limit, so that's why it worked.
 
 Bumped the descriptor limit to 1024 for the 3 types we need.

Keenuts commented:
Khronos does not want to push vendor extensions, meaning no DDS extension in samples, meaning no support for it for cgltf (Pr pending, but no info yet).
Fixed the sample so it can load an untained intel sponza scene.
PNG loading is way slower (68s vs 0.7s to load textures). KTX2 was a bit slower as it does transcoding, but we can stash a DDS into KTX2 and detect it using the key-value storage in the file. That might be the best option yet to have a fast loading.

vettoreldaniele commented on projects/28_gltf/main.cpp:
```diff
@@ -0,0 +1,837 @@
+        scene.ambient                    = float4(0.3f);
+        scene.cameraViewProjectionMatrix = mCamera.GetViewProjectionMatrix();
+        scene.lightPosition              = float4(mLightPosition, 0);
+        scene.eyePosition                = glm::float4(mCamera.GetEyePosition(), 0.f);
```
can drop the `glm::` here

vettoreldaniele commented on projects/28_gltf/main.cpp:
```diff
@@ -0,0 +1,837 @@
+        constexpr size_t                                    TEXTURE_COUNT    = 3;
+        constexpr size_t                                    DESCRIPTOR_COUNT = 1 + TEXTURE_COUNT * 2 /* uniform + 3 * (sampler + texture) */;
+        std::array<grfx::WriteDescriptor, DESCRIPTOR_COUNT> write;
+        std::memset(write.data(), 0, sizeof(write[0]) * write.size());
```
This is already initialized through the default constructor. All `grfx::` types initialize to something sensible

Keenuts replied:
Ah thanks, good to know.

vettoreldaniele commented on projects/28_gltf/main.cpp:
```diff
@@ -0,0 +1,837 @@
+        object.pUniformBuffer->CopyFromSource(sizeof(scene), &scene);
+    }
+
+    {
```
Do we need to update the descriptors at every frame? I don't see anything that changes across frames.

Keenuts replied:
We don't for now (no camera movement), but we should add an orbital camera right after this is implemented.

vettoreldaniele commented on projects/28_gltf/main.cpp:
```diff
@@ -0,0 +1,837 @@
+    PerspCamera                  mCamera;
+    float3                       mLightPosition = float3(10, 100, 10);
+
+    RenderList             mRenderList;
```
This is never used. I suppose you wanted to use this to sort by material and reduce the state switching?

Keenuts replied:
Oh yes, thanks, yes iterated a bit around this, forgot to clean this up.

vettoreldaniele reviewed (dismissed):
Do you plan on having some of the loading functionality moved to the `grfx::` namespace as a helper?

Keenuts commented on .gitmodules:
```diff
@@ -40,3 +40,7 @@
+[submodule "third_party/cgltf"]
+	path = third_party/cgltf
+	url = https://github.com/jkuhlmann/cgltf.git
+  branch = dds
```
forgot to remove this line, thanks!

Keenuts commented on assets/basic/shaders/PbrMetallicRoughness.hlsl:
```diff
@@ -0,0 +1,148 @@
+    if (albedo.a < 0.8f) {
+      discard;
+    }
+    const float roughness = MetalRoughness.Sample(MetalRoughnessSampler, input.uv).g;
```
No, the green one. The difference between the altimeter texture and the sponza scene is the red channel being either all-white, or all-black. But on both cases, green is roughness, blue metalness.
(looking at the asset is looks like red is used, but it's just because of the default to 1.0)

Keenuts commented on projects/28_gltf/main.cpp:
```diff
@@ -0,0 +1,837 @@
+    settings.window.width               = 1920;
+    settings.window.height              = 1080;
+    settings.grfx.api                   = kApi;
+    settings.grfx.enableDebug           = true;
```
Yes, good point.

Keenuts commented on projects/28_gltf/main.cpp:
```diff
@@ -0,0 +1,837 @@
+    }
+
+    if (material.normal_texture.texture == nullptr) {
+        LoadTexture(ColorToBitmap(float3(0.f, 1.f, 0.f)), pQueue, &pOutput->textures[1]);
```
Was supposed to be `(0, 0, 1)` fixed! The goal was to have a neutral normal (tangent space) in case we don't have more info than the triangle's normal.
Setting it to `0, 0, 0` would not look.
3 screenshots: with normal map, without (0, 0, 1), and without (0, 0, 0):
![normal](https://user-images.githubusercontent.com/4282252/227889958-c302541f-e54c-4f8a-91d5-9242878e9baf.png)
![no-normal](https://user-images.githubusercontent.com/4282252/227889979-4f2cec9c-3990-4092-be32-b3c4ffda73f5.png)
![all-zero](https://user-images.githubusercontent.com/4282252/227889995-66e08597-32bc-426e-97ec-d89fbab57978.png)

Keenuts commented on projects/28_gltf/main.cpp:
```diff
@@ -0,0 +1,837 @@
+    }
+
+    {
+        // FIXME: this assumes we have only PBR, and with 3 materials per texture. Needs to be revisited.
```
Yes, thanks

Keenuts commented:
> Do you plan on having some of the loading functionality moved to the `grfx::` namespace as a helper?

Yes, but not proactively. The idea is to see if there is some usage, how it's used, and from that derive an API/framework for the scene description.

wangra-google commented on assets/basic/shaders/PbrMetallicRoughness.hlsl:
```diff
@@ -0,0 +1,148 @@
+    const float3 specularBRDF = (F * D * G) / max(0.00001, 4.0 * cosLi * cosLo);
+    const float3 Co = (diffuseBRDF + specularBRDF) * Lrad * cosLi + 0.3f * albedo.rgb;
+
+    return float4(saturate(Co), 1);
```
Any reason you want to clamp the range to [0,1]?

wangra-google replied:
it would be wrong to normalize the output this way for pbr, should use tonemapping to convert from hdr to sdr.

Keenuts commented on projects/28_gltf/main.cpp:
```diff
@@ -0,0 +1,837 @@
+    PPX_ASSERT_MSG(*ppPosition != nullptr && *ppUv != nullptr && *ppNormal != nullptr && *ppTangent != nullptr, "For now, only supports model with position, normal, tangent and UV attributes");
+}
+
+void ProjApp::LoadPrimitive(const cgltf_primitive& primitive, grfx::BufferPtr pStagingBuffer, grfx::Queue* pQueue, Primitive* pOutput) const
```
added a comment on the prototype definition.

Keenuts commented on projects/28_gltf/main.cpp:
```diff
@@ -0,0 +1,837 @@
+
+    struct Object
+    {
+        float4x4                model;
```
`modelMatrix` to match the other Scene struct.

Keenuts commented on projects/28_gltf/main.cpp:
```diff
@@ -0,0 +1,837 @@
+            const size_t primitive_index = primitiveToIndex.at(&node.mesh->primitives[j]);
+            // FIXME: support meshes with no material.
+            // For now, assign the first available material.
+            PPX_ASSERT_MSG(pMaterials->size() != 0, "Doesn't support GLTF files with no materials.");
```
Not sure that would be the best location, as it's not an issue with LoadMaterial, but with the mesh creation step, where we should lazily create some material if none is defined no?

Keenuts commented on assets/basic/shaders/PbrMetallicRoughness.hlsl:
```diff
@@ -0,0 +1,148 @@
+    const float3 kD = lerp(float3(1, 1, 1) - F, float3(0, 0, 0), metalness);
+    const float3 diffuseBRDF = kD * albedo.rgb;
+    const float3 specularBRDF = (F * D * G) / max(0.00001, 4.0 * cosLi * cosLo);
+    const float3 Co = (diffuseBRDF + specularBRDF) * Lrad * cosLi + 0.3f * albedo.rgb;
```
Replaced 0.3 with Scene.Ambient, nice catch forgot this one

Keenuts commented on projects/28_gltf/main.cpp:
```diff
@@ -0,0 +1,837 @@
+
+    // Update uniform buffers
+    for (auto& object : mObjects) {
+        // Draw uniform buffers
```
of course, also duplicate of the comment above, removed!

wangra-google commented on assets/basic/shaders/PbrMetallicRoughness.hlsl:
```diff
@@ -0,0 +1,149 @@
+    const float3 specularBRDF = (F * D * G) / max(0.00001, 4.0 * cosLi * cosLo);
+    const float3 Co = (diffuseBRDF + specularBRDF) * Lrad * cosLi + Scene.Ambient * albedo.rgb;
+
+    return float4(Co, 1);
```
you will need tonemapping to convert from hdr to sdr. rainhard should be fine.

Keenuts replied:
one detail: this "PBR" shader is really just so we can visualize the pbr material. It's not correct.

wangra-google replied:
just one more line for rainhard: Co = Co/(Co+float3(1,1,1));

Keenuts replied:
done, thanks!
I should invest in an HDR screen one day :smile:

Keenuts commented:
Sorry for the force-push, messed-up while rebasinb. Should be fine now.
